dart/collision/CollisionGroup.hpp

```diff
@@ -265,6 +376,146 @@ class CollisionGroup
   // original and copy are not guranteed to be the same as we copy std::map
   // (e.g., by world cloning).
 
+
+private:
+
+  /// This class watches when ShapeFrames get deleted so that they can be safely
+  /// removes from the CollisionGroup. We cannot have a weak_ptr to a ShapeFrame
+  /// because some are managed by std::shared_ptr while others are managed by
+  /// NodePtr.
+  ///
+  /// If we don't carefully track when each ShapeFrame gets destructed, then we
+  /// can land in a situation where a ShapeFrame gets removed from a BodyNode,
+  /// deallocated, and then a new ShapeFrame is allocated in the memory address
+  /// of the old one. This can lead to invalid memory accesses if we neglect to
+  /// correctly clean up our references to deleted ShapeFrames.
+  class ShapeFrameObserver final : public common::Observer
+  {
+  public:
+
+    /// Add a shape frame to this observer
+    void addShapeFrame(const dynamics::ShapeFrame* shapeFrame);
+
+    /// Remove a shape frame from this observer
+    void removeShapeFrame(const dynamics::ShapeFrame* shapeFrame);
+
+    /// Remove all shape frames from this observer
+    void removeAllShapeFrames();
+
+    /// Whenever an observed shape frame gets deleted, its pointer will be added
+    /// to this set. The next time the collision group updates, it will check this
+    /// set and wipe away any references to the pointers in this set.
+    std::unordered_set<const dynamics::ShapeFrame*> mDeletedFrames;
+
+  protected:
+
+    /// This will be called each time an observed shape frame is deleted.
+    void handleDestructionNotification(const common::Subject* subject) override;
+
+  private:
+
+    /// A map from a subject pointer to its corresponding ShapeFrame pointer.
+    /// This needs to be stored because by the time a Subject is being
+    /// destructed, it can no longer be cast back to its ShapeFrame.
+    std::unordered_map<const common::Subject*,
+                       const dynamics::ShapeFrame*> mMap;
+
+  };
+
+  /// \private Implementation of addShapeFrame. The source argument tells us
+  /// whether this ShapeFrame is being requested explicitly by the user or
+  /// implicitly through a BodyNode, Skeleton, or other CollisionGroup.
+  ObjectInfo* _addShapeFrameImpl(
```

Comment:
I think the convention varies a bit throughout the source code, but that's probably just because I have a long-time habit of decorating private things with underscores (I like to make it obvious from the function name that a function isn't part of the public API). I can remove it if it's bothersome.